### PR TITLE
Add MultiSearchTemplate support to High Level Rest client

### PR DESCRIPTION
**Pull request by markharwood**
Note some changes to core classes were required to support tests:

* Added SearchType.currentlySupported to list only the usable options
* Changed SearchResponse.Clusters constructor to public
* Added SearchRequest.DEFAULT_BATCHED_REDUCE_SIZE constant
* Added missing "tookInMillis" to MultiSearchTemplateResponse
* Added SearchTemplateResponse toString() for comparisons in tests

### CODE REVIEW

File: client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java [hunk 1]
```diff
@@ -1229,7 +1235,103 @@ public void testRenderSearchTemplate() throws Exception {
...
+                sr.source(new SearchSourceBuilder());
+            }
+        }        
+        // Compare the deserialized request object with the original request 
+        assertEquals(multiSearchTemplateRequest, deser);
+        
+        // Finally, serialize the deserialized request to compare JSON equivalence (in case Object.equals() fails to reveal a discrepancy)
+        String reserJsonRequestString = new String(MultiSearchTemplateRequest.writeMultiLineFormat(deser, XContentType.JSON.xContent()),
+                StandardCharsets.UTF_8);
+        assertEquals(jsonRequestString, reserJsonRequestString);
```
Comment (javanna) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java]: I think that the multiline format should be tested in elasticsearch given that the code belongs there? Something along the lines of `MultiSearchRequestTests#testMultiLineSerialization`?

File: client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java [hunk 2]
```diff
@@ -1834,7 +1936,14 @@ private static void setRandomIndicesOptions(Consumer<IndicesOptions> setter, Sup
...
         if (randomBoolean()) {
-            setter.accept(IndicesOptions.fromOptions(randomBoolean(), randomBoolean(), randomBoolean(), randomBoolean()));
+            // This caused failures in testMultiSearchTemplate's request equality assertion: 
+            // setter.accept(IndicesOptions.fromOptions(randomBoolean(), randomBoolean(), randomBoolean(), randomBoolean()));
+            
+            // See https://github.com/elastic/elasticsearch/issues/20105#issuecomment-265692775 
+            // The above notes the discrepancy between REST-configurable params and those in the Java client. 
+            // It proposes the solution below:
+            setter.accept(IndicesOptions.fromOptions(randomBoolean(), randomBoolean(), randomBoolean(), randomBoolean(), 
+                    SearchRequest.DEFAULT_INDICES_OPTIONS));
```
Comment (javanna) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java]: good point, but I don't think that using the default options for search works for all the other API? At least not in theory. maybe we should allow to provide the default ones for each API?

File: modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/MultiSearchTemplateResponse.java [hunk 3]
```diff
@@ -143,11 +170,15 @@ public void writeTo(StreamOutput out) throws IOException {
...
         }
+        if (out.getVersion().onOrAfter(Version.V_7_0_0_alpha1)) {
+            out.writeVLong(tookInMillis);
+        }
     }
 
     @Override
     public XContentBuilder toXContent(XContentBuilder builder, ToXContent.Params params) throws IOException {
         builder.startObject();
+        builder.field("took", tookInMillis);
```
Comment (javanna) [inline modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/MultiSearchTemplateResponse.java]: I guess this should have been added as part of #23767 . Can we make this change in a separate PR please? That way it will have a bit more visibility?

File: modules/lang-mustache/src/test/java/org/elasticsearch/script/mustache/MultiSearchTemplateResponseTests.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+import java.io.IOException;
+
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.nullValue;
+
+// Largely a copy of MultiSearchResponseTests - because multi template results are 
+// just wrapper objects for regular multi search results
+public class MultiSearchTemplateResponseTests extends ESTestCase {
```
Comment (javanna) [inline modules/lang-mustache/src/test/java/org/elasticsearch/script/mustache/MultiSearchTemplateResponseTests.java]: would it be possible to extend from AbstractXContentTestCase?

File: server/src/main/java/org/elasticsearch/action/search/SearchType.java [hunk 5]
```diff
@@ -50,6 +50,11 @@
      */
     public static final SearchType DEFAULT = QUERY_THEN_FETCH;
 
+    /**
+     * Non-deprecated types
+     */
+    public static final SearchType [] CURRENTLY_SUPPORTED = {QUERY_THEN_FETCH, DFS_QUERY_THEN_FETCH};
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/action/search/SearchType.java]: curious about this: what problems were there with deprecated values? Given that it's only used in tests, maybe this could be only in the test where it's needed?

**Review by javanna — CHANGES_REQUESTED** [review]
Left a couple of comments, thanks @markharwood looks good. Can you also add docs please?

File: modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/MultiSearchTemplateResponse.java (hunk 3 above)
Comment (markharwood) [reply modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/MultiSearchTemplateResponse.java]: OK - that issue/PR will have to be pushed before this one because the fromXContent part of this PR was failing due to the missing `tookInMillis` property.

File: server/src/main/java/org/elasticsearch/action/search/SearchType.java (hunk 5 above)
Comment (markharwood) [reply server/src/main/java/org/elasticsearch/action/search/SearchType.java]: The test was randomly picking from enum's values but the "fromString" code in the same enum would barf on one of those values.

>Given that it's only used in tests, maybe this could be only in the test where it's needed?

I thought there would be less chance of getting it wrong in future if the enum where the values are decoded also listed the acceptable values it can decode.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java (hunk 1 above)
Comment (markharwood) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java]: > I think that the multiline format should be tested in elasticsearch given that the code belongs there?

Perhaps not core - MultiSearchTemplateRequest is in a module because of the mustache dependency. Maybe a new class in /:modules:lang-mustache/src/test/java/org.elasticsearch.script.mustache ?

File: client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java (hunk 2 above)
Comment (markharwood) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java]: I found this "options" part of the code pretty confusing to be honest. Could do with a quick chat about what's going on here.

File: modules/lang-mustache/src/test/java/org/elasticsearch/script/mustache/MultiSearchTemplateResponseTests.java (hunk 4 above)
Comment (markharwood) [reply modules/lang-mustache/src/test/java/org/elasticsearch/script/mustache/MultiSearchTemplateResponseTests.java]: Not sure that would work - AbstractXContentTestCase tests for equality using toString() on the response object and any exceptions in the server-side instance don't compare exactly with the re-materialized exceptions from Xcontent. Likely this is why the MultiSearchResponseTests (from which I copied this test) does not use this base class.

Comment (javanna) [reply modules/lang-mustache/src/test/java/org/elasticsearch/script/mustache/MultiSearchTemplateResponseTests.java]: you can override assertEqualInstances

Comment (javanna) [reply modules/lang-mustache/src/test/java/org/elasticsearch/script/mustache/MultiSearchTemplateResponseTests.java]: MultiSearchResponseTests was written before the test base classes were made more flexible, we should probably migrate that too if it works well for your case

File: client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java (hunk 1 above)
Comment (javanna) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java]: yea I mean where the request sits.

Comment (javanna) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java]: to clarify, I would leave the initial part and test here only the result of `RequestConverters#multiSearchTemplate`, and would move the multi line serialization roundtrips to another test to be placed where the request lives.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java (hunk 2 above)
Comment (javanna) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java]: `IndicesOptions` holds some options that can be set at REST (allow_no_indices, Ignore_unavailable, expand_wildcards) and some others that describe the behaviour of an API, which are hardcoded (on purpose). Unfortunately the transport client allows to set also the latter, with catastrophic results though :)

The point is that, for instance, the search API cannot work against closed indices, hence forbid closed indices is set to true in the default search api indices options, which means that closed indices are considered "unavailable". The default indices options matter as they hold default values, but also some values that cannot be overridden (unless you do it from the transport client at your own risk).

This is not a problem in the high-level REST client, as you can potentially set values like forbidClosedIndices etc. but they will be ignored, only the ones supported at REST will be converted to query_string params. That is what we check in these tests. I suspect that this change is needed because of your request comparison which should not be here but rather in the specific test around multiline serialization.

**Comment by markharwood** [comment]
Happy with the latest changes, @javanna ?

File: modules/lang-mustache/src/test/java/org/elasticsearch/script/mustache/MultiSearchTemplateResponseTests.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    @Override
+    protected MultiSearchTemplateResponse doParseInstance(XContentParser parser) throws IOException {
+        return MultiSearchTemplateResponse.fromXContext(parser);
+    }
+
+    @Override
+    protected boolean supportsUnknownFields() {
+        return false;
```
Comment (javanna) [inline modules/lang-mustache/src/test/java/org/elasticsearch/script/mustache/MultiSearchTemplateResponseTests.java]: ideally, you would leave this to true and exclude some of the paths where injecting random fields is not supported, I can imagine that the top-level is problematic here and the _source part as well. You can override `getRandomFieldsExcludeFilter`.

File: modules/lang-mustache/src/test/java/org/elasticsearch/script/mustache/MultiSearchTemplateResponseTests.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    @Override
+    protected boolean assertToXContentEquivalence() {
+        return false;
```
Comment (javanna) [inline modules/lang-mustache/src/test/java/org/elasticsearch/script/mustache/MultiSearchTemplateResponseTests.java]: what made you return false here? checking whether this is expected or not. may have to do with exceptions which are parsed differently compared to their original representation. In that case, we could have an ordinary test without exceptions and a variation of it with exceptions where we disabled this check. We do this already in `ListTasksResponseTests`

File: server/src/test/java/org/elasticsearch/action/search/MultiSearchResponseTests.java [hunk 8]
```diff
@@ -83,4 +55,36 @@ private static MultiSearchResponse createTestInstance() {
...
+
+    @Override
+    protected boolean supportsUnknownFields() {
+        return false;
+    }
+
+    @Override
+    protected boolean assertToXContentEquivalence() {
+        return false;
+    }
```
Comment (javanna) [inline server/src/test/java/org/elasticsearch/action/search/MultiSearchResponseTests.java]: same comments as above, thanks for converting this test though!!!

**Review by javanna — CHANGES_REQUESTED** [review]
left a couple more comments, mainly on testing, thanks @markharwood

File: modules/lang-mustache/src/test/java/org/elasticsearch/script/mustache/MultiSearchTemplateResponseTests.java (hunk 6 above)
Comment (markharwood) [reply modules/lang-mustache/src/test/java/org/elasticsearch/script/mustache/MultiSearchTemplateResponseTests.java]: I tried it and the debugger showed me it failing on an unknown random field name (something like "XCZXFDVC"). Is the end goal here to ensure strictness is being enforced (the test correctly rejects and reports any random noise in the xcontent fields) or that tolerance is applied (the rest client will silently tolerate unknown additions)?

Comment (javanna) [reply modules/lang-mustache/src/test/java/org/elasticsearch/script/mustache/MultiSearchTemplateResponseTests.java]: that tolerance is applied, it's a way to test forward compatibility of the client.

Comment (javanna) [reply modules/lang-mustache/src/test/java/org/elasticsearch/script/mustache/MultiSearchTemplateResponseTests.java]: obviously there are places where we injecting random fields makes no sense, like _source etc

Comment (markharwood) [reply modules/lang-mustache/src/test/java/org/elasticsearch/script/mustache/MultiSearchTemplateResponseTests.java]: Makes sense. I don't think `getRandomFieldsExcludeFilter ` will be enough and the parsing code in  `MultiSearchResponse.Item#itemFromXContent` will need changing. Currently it has branching logic like this: 

	 if ("error".equals(fieldName)) {
		item = new Item(null, ElasticsearchException.failureFromXContent(parser));
	} else  ...{
		// assume non-error and parse the hits etc
		item = new Item(SearchResponse.innerFromXContent(parser), null);

The `getRandomFieldsExcludeFilter` only lets me filter on the name but here I would need to filter if the object context being injected into already contains an `error` field as the code assumes that errors are the only field in an error response.

File: server/src/test/java/org/elasticsearch/action/search/MultiSearchResponseTests.java [hunk 9]
```diff
@@ -83,4 +78,53 @@ private static MultiSearchResponse createTestInstance() {
...
+
+    @Override
+    protected boolean supportsUnknownFields() {
+        return true;
+    }
+
+    @Override
+    protected Predicate<String> getRandomFieldsExcludeFilter() {
+        return field -> field.startsWith("responses");
```
Comment (javanna) [inline server/src/test/java/org/elasticsearch/action/search/MultiSearchResponseTests.java]: with the recent changes, I think that you only need the filter above when testing with failures. here and in the other test.

**Review by javanna — APPROVED** [review]
LGTM besides the small comment I left on testing. thanks @markharwood !

**Comment by markharwood** [comment]
run gradle build tests
